### PR TITLE
feat: introduce `wdl-lsp` crate.

### PR DESCRIPTION
**Pull request by peterhuene**
This commit introduces the `wdl-lsp` crate.

The new crate is responsible for implementing an LSP server that can be used in compatible code editors.

The LSP server currently implements the following:

* Document diagnostics.
* Workspace diagnostics.
* Incremental document changes.
* Monitoring the workspace for file/directory changes.
* Support for multiple workspace folders.
* Notifying the client regarding work progress for long-running tasks.

To assist the implementation of the LSP, much if `wdl-analysis` was rewritten.

`AnalysisEngine` was renamed to simply `Analyzer` and much of its implementation was moved to an internal "analysis queue" type.

The progress callback to the analyzer was made async so that an LSP server may notify the client regarding the progress.

Before submitting this PR, please make sure:

- [x] You have added a few sentences describing the PR here.
- [x] You have added yourself or the appropriate individual as the assignee.
- [x] You have added at least one relevant code reviewer to the PR.
- [x] Your code builds clean without any errors or warnings.
- [x] You have added tests (when appropriate).
- [x] You have updated the README or other documentation to account for these
      changes (when appropriate).
- [x] You have added an entry to the relevant `CHANGELOG.md` (see
      ["keep a changelog"] for more information).
- [x] Your commit messages follow the [conventional commit] style.

[conventional commit]: https://www.conventionalcommits.org/en/v1.0.0/#summary
["keep a changelog"]: https://keepachangelog.com/en/1.0.0/

### CODE REVIEW

File: wdl-analysis/src/analyzer.rs [hunk 1]
```diff
@@ -0,0 +1,478 @@
...
+pub struct AnalysisResult {
+    /// The URI of the analyzed document.
+    uri: Arc<Url>,
+    /// The root node of the document.
+    ///
+    /// This is `None` if the document failed to be read.
+    root: Option<GreenNode>,
+    /// The line index used to map line/column offsets to byte offsets and vice
+    /// versa.
+    lines: Option<Arc<LineIndex>>,
```
Comment (claymcleod) [inline wdl-analysis/src/analyzer.rs:73]: When is this `None`?

File: wdl-analysis/src/analyzer.rs [hunk 2]
```diff
@@ -0,0 +1,478 @@
...
+}
+
+/// Represents a position in a document's source.
+#[derive(Debug, Eq, PartialEq, Ord, PartialOrd, Copy, Clone, Default)]
+pub struct SourcePosition {
+    /// Line position in a document (zero-based).
+    pub line: u32,
+    /// Character offset on a line in a document (zero-based). The meaning of
+    /// this offset is determined by the position encoding.
+    pub character: u32,
```
Comment (claymcleod) [inline wdl-analysis/src/analyzer.rs:165]: Is this better as `col`? Or am I misunderstanding.

Comment (claymcleod) [reply wdl-analysis/src/analyzer.rs:165]: Perhaps this is tied to the LSP specification's language—I didn't check but, now that I'm reading further, it looks to be used there.

File: wdl-analysis/src/analyzer.rs [hunk 3]
```diff
@@ -0,0 +1,478 @@
...
+        &self.diagnostics
+    }
+
+    /// Gets the scope of the analyzed document.
+    pub fn scope(&self) -> &Arc<DocumentScope> {
+        &self.scope
+    }
+}
+
+/// Represents a position in a document's source.
```
Comment (claymcleod) [inline wdl-analysis/src/analyzer.rs:158]: Is it worth noting that the order of the fields is important here?

File: wdl-lsp/CHANGELOG.md [hunk 4]
```diff
@@ -0,0 +1,12 @@
+# Changelog
+
+All notable changes to this project will be documented in this file.
+
+The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.1.0/),
+and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).
+
+## Unchanged
```
Comment (claymcleod) [inline wdl-lsp/CHANGELOG.md:8]:
```suggestion
## Unreleased
```

File: wdl-lsp/src/lib.rs [hunk 5]
```diff
@@ -0,0 +1,485 @@
...
+                    },
+                    move || validator(lint),
+                    move |uri| changes(&analyzer_workspace, uri),
+                ),
+
+                workspace,
+            }
+        });
+
+        let stdin = tokio::io::stdin();
```
Comment (claymcleod) [inline wdl-lsp/src/lib.rs:124]: I'm not sure if we want to give some thought as to how to convey the communication method (https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#implementationConsiderations)? Here, we've hardcoded the `stdio` method—I know that's what a lot of implementations do, but do we want to make that more apparent in the initialization somehow?

Regardless, I still recommend `sprocket` eventually require the `--stdio` flag to make it explicit, even if there is only one option.

File: wdl-lsp/src/lib.rs [hunk 6]
```diff
@@ -0,0 +1,485 @@
...
+        {
+            return None;
+        }
+
+        self.client
+            .send_notification::<Progress>(ProgressParams {
+                token: NumberOrString::String(token.clone()),
+                value: ProgressParamsValue::WorkDone(WorkDoneProgress::Begin(
+                    WorkDoneProgressBegin {
+                        title: "Sprocket".to_string(),
```
Comment (claymcleod) [inline wdl-lsp/src/lib.rs:177]: Since this technically _isn't_ sprocket yet, how do we want to handle this?

File: wdl-lsp/src/proto.rs [hunk 7]
```diff
@@ -0,0 +1,85 @@
...
+                message: label.message().to_string(),
+            })
+        })
+        .collect::<Result<_>>()?;
+
+    Ok(Diagnostic::new(
+        range.unwrap_or_default(),
+        Some(severity),
+        code,
+        Some("sprocket".to_string()),
```
Comment (claymcleod) [inline wdl-lsp/src/proto.rs:80]: Same thought here.

**Review by claymcleod — APPROVED** [review]
Looks really good. I don't really have any notes beyond what I put above.

File: wdl-analysis/CHANGELOG.md [hunk 8]
```diff
@@ -5,6 +5,12 @@ All notable changes to this project will be documented in this file.
 The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.1.0/),
 and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).
 
+## Unchanged
```
Comment (a-frantz) [inline wdl-analysis/CHANGELOG.md:8]:
```suggestion
## Unreleased
```

File: wdl-lsp/src/lib.rs (hunk 6 above)
Comment (a-frantz) [reply wdl-lsp/src/lib.rs:177]: I think `wdl-lsp` is more appropriate

Comment (a-frantz) [reply wdl-lsp/src/lib.rs:177]: Or the more flexible `env!("CARGO_CRATE_NAME")`? (copied from a bit further down)

File: wdl-analysis/src/analyzer.rs (hunk 1 above)
Comment (peterhuene) [reply wdl-analysis/src/analyzer.rs:73]: It's `None` when `root` is also `None`, indicating that there was an error fetching the document (i.e. doesn't exist on disk or HTTP error).

These three fields (`root`, `lines`, and `error`) are probably better encapsulated in an enum. I'll write that down for a follow-up fix.

File: wdl-analysis/src/analyzer.rs (hunk 2 above)
Comment (peterhuene) [reply wdl-analysis/src/analyzer.rs:165]: Indeed, it was chosen to match the LSP type; effectively the same thing.

File: wdl-analysis/src/analyzer.rs (hunk 3 above)
Comment (peterhuene) [reply wdl-analysis/src/analyzer.rs:158]: Which order are you referring to?

**Review by adthrasher — APPROVED** [review]

File: wdl-lsp/src/lib.rs (hunk 5 above)
Comment (peterhuene) [reply wdl-lsp/src/lib.rs:124]: The forthcoming `sprocket analyzer` command implementation does take a `--stdio` option.

I plan on extending the construction of the server to take the transport when I implement tests, so we can spawn the server on as a tokio task and hand it an in-memory communication channel.

File: wdl-lsp/src/lib.rs (hunk 6 above)
Comment (peterhuene) [reply wdl-lsp/src/lib.rs:177]: Agreed, this should be coming from the sever options as a "display name" for the server.

We'll default to the crate name if not specified.

File: wdl-analysis/src/analyzer.rs (hunk 1 above)
Comment (peterhuene) [reply wdl-analysis/src/analyzer.rs]: I pushed up a fix for this.

File: wdl-analysis/src/analyzer.rs (hunk 3 above)
Comment (claymcleod) [reply wdl-analysis/src/analyzer.rs:184]: Here, I was referring to the sort order the is implied by the order of the members. That order is depended on later on in the code when comparisons are done.

It's arguably obvious if you understand how `#[derive(Ord, PartialOrd)]` work, but sometime you just define the order such that they have _some_ order (deterministic sorting, for example) whereas sometimes you have an order that has meaning outside of an arbitrary order. If it were me, I'd just note that the ordering of the members matter in the comment above to make that abundantly apparent.

Comment (peterhuene) [reply wdl-analysis/src/analyzer.rs:184]: I'll add that note.

**Review by claymcleod — APPROVED** [review]

**Pull request merged**